File: blind75/intervals/insert_interval.py
from typing import List

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []

        for i in range(len(intervals)):
            if intervals[i][1] < newInterval[0]:
                res.append(intervals[i])
            elif intervals[i][0] > newInterval[1]:
                res.append(newInterval)
                return res + intervals[i:]
            else:
                newInterval = [
                    min(newInterval[0], intervals[i][0]),
                    max(newInterval[1], intervals[i][1])
                ]
        res.append(newInterval)
        return res


    def insert2(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        # *intervals already sorted by start in problem desc* 
        if not intervals:
            return [newInterval]
        res = []
        i = 0
        while i < len(intervals) and intervals[i][1] < newInterval[0]:
            res.append(intervals[i])
            i += 1 
        # merge intervals as long as new interval doesn't 
        # start before itv ends or ends before itv starts
        while i < len(intervals) and (intervals[i][1] >= newInterval[0] and intervals[i][0] <= newInterval[1]):
            newInterval[0] = min(intervals[i][0], newInterval[0])
            newInterval[1] = max(intervals[i][1], newInterval[1])
            i += 1
        res.append(newInterval)
        while i < len(intervals):
            res.append(intervals[i])
            i += 1
        return res

File: blind75/intervals/insert_interval.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new ArrayList<>();

        for (int[] itv : intervals) {
            if (newInterval == null || itv[1] < newInterval[0]) {
                res.add(itv);
            } else if (itv[0] > newInterval[1]) {
                res.add(newInterval);
                res.add(itv);
                newInterval = null;
            } else {
                newInterval[0] = Math.min(itv[0], newInterval[0]);
                newInterval[1] = Math.max(itv[1], newInterval[1]);
            }
        }

        if (newInterval != null) {
            res.add(newInterval);
        }

        return res.toArray(new int[res.size()][]);
    }
}